bletl/growth.py

```diff
@@ -500,3 +456,69 @@ def fit_mu_t(
         result.sample(draws=mcmc_samples)
 
     return result
+
+
+def detect_switchpoints(
+    switchpoint_prob: float,
+    t_data: typing.Sequence[float],
+    pmodel: pm.Model,
+    theta_map: typing.Dict[str, numpy.ndarray],
+) -> typing.Dict[float, str]:
+    """Helper function to detect switchpoints from a fitted random walk.
+
+    Parameters
+    ----------
+    switchpoint_prob
+        Probability threshold for detecting switchpoints.
+        Random walk innovations with a prior probability less than this
+        will be classified as switchpoints.
+    t_data
+        Time values corresponding to the random walk steps.
+    pmodel
+        The PyMC model containing `"mu_t*"` random walks.
+    theta_map
+        MAP estimate of the model.
+
+    Returns
+    -------
+    switchpoints
+        Dictionary of switchpoints with
+        keys being the time point and
+        values `"detected"`.
+    """
+    # first CDF values at all mu_t elements
+    cdf_evals = []
+    for rvname in sorted(theta_map.keys()):
+        if rvname not in pmodel.named_vars:
+            continue
+        # The random walk may be split in multiple segments.
+        # We can identify a segment from the RVOp type that created it.
+        rv = pmodel[rvname]
+        if rv.owner is None:
+            continue
+        if isinstance(rv.owner.op, pm.RandomWalk.rv_type):
+            # Get a handle on the innovation dist
+            innov_dist = rv.owner.inputs[1]
+            # and calculate the innovations from the MAP estimate
+            innov = numpy.diff(theta_map[rvname])
+            # so we can calculate the logcdf and cdf evaluations.
+            logcdfs = pm.logcdf(innov_dist, innov).eval()
+            # Prepend a 0.5 because the init dist does not count into the diff.
+            cdf_evals += [0.5, *numpy.exp(logcdfs)]
+    cdf_evals = numpy.array(cdf_evals)
+    if len(cdf_evals) != len(t_data) - 1:
+        raise Exception(
+            f"Failed to find all random walk segments. Found {len(cdf_evals)}, expected {len(t_data) - 1}."
+        )
+    # filter for the elements that lie outside of the [0.005, 0.995] interval
+    significance_mask = numpy.logical_or(
+        cdf_evals < (switchpoint_prob / 2),
+        cdf_evals > (1 - switchpoint_prob / 2),
+    )
+    # add these autodetected timepoints to the switchpoints-dict
+    # (ignore the first timepoint)
+    switchpoints = {}
+    for t_switch, is_switchpoint in zip(t_data, significance_mask[1:]):
```

Comment:
shouldn't you also mask t_data by [1:] here?

Reply:
This is a matter of how the switchpoint time is defined.
If anything it should be shifted by one more element, because in the notebook the detected switchpoint is actually one element too early?

This is why I added the test case where it finds exactly the right time value as expected.

Reply:
With the new changes, the zip now iterates over two equally sized collections and no longer slices anything.